### PR TITLE
New post about RESTful APIs

### PR DESCRIPTION
**Pull request by danielwpz**
I have read about a very clear and helpful article discussing how to design well-formed RESTful APIs. However, the original article is written in Chinese. So I translated that one and wanna to post it in order to share with everyone. :)

### CODE REVIEW

**Comment by zcohn** [comment]
There are a few spelling or whitespace errors here to fix (thesedays, protocal, infomation, etc.) might want to put this into a text editor with spellcheck. Although I see from the license at the end that you aren't modifying the original, it does feels like more of a description of idempotence (and a few other aspects) would be helpful to clarify, perhaps before the body of the translated work.

**Comment by danielwpz** [comment]
Thanks, Zach. I tried to fix the spelling errors and move that license part before the body.

File: _posts/2016-06-24-restful-api-design-principles.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+For example:    
+* ```?limit=10``` : Return only 10 records.
+* ```?page=2&per_page=10``` : Return the records in the 2nd page, if each page contains 10 records.
+
+The design of parameters could be redundant, which means both ```/zoo/ID/animals``` and ```/animals?zoo_id=ID``` refers to the same resource.    
+
+## Status Codes
+
+A respond from the server could have no content, however, each response MUST contain a **status code**.    
+* ```200 OK``` : The resource that client asked is successfully returned. *And also, the request caused 200 is idempotent.*
```
Comment (jkenny) [inline _posts/2016-06-24-restful-api-design-principles.md]: "And also, the request is idempotent."

File: _posts/2016-06-24-restful-api-design-principles.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+For example:    
+* ```?limit=10``` : Return only 10 records.
+* ```?page=2&per_page=10``` : Return the records in the 2nd page, if each page contains 10 records.
+
+The design of parameters could be redundant, which means both ```/zoo/ID/animals``` and ```/animals?zoo_id=ID``` refers to the same resource.    
+
+## Status Codes
+
+A respond from the server could have no content, however, each response MUST contain a **status code**.
```
Comment (jkenny) [inline _posts/2016-06-24-restful-api-design-principles.md]: s/respond/response
s/could/can

File: _posts/2016-06-24-restful-api-design-principles.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+The design of parameters could be redundant, which means both ```/zoo/ID/animals``` and ```/animals?zoo_id=ID``` refers to the same resource.    
+
+## Status Codes
+
+A respond from the server could have no content, however, each response MUST contain a **status code**.    
+* ```200 OK``` : The resource that client asked is successfully returned. *And also, the request caused 200 is idempotent.*
+* ```201 CREATED``` : Client has successfully created or updated a resource.
+* ```202 ACCEPTED``` : The request has been accepted but the server chooses to process it later.
+* ```204 NO CONTENT``` : The resource has been successfully deleted.
+* ```400 INVALID REQUEST``` : There is some error in the request sent by the client. And the server hasn't done anything with the request.  *And also, the request caused 400 is idempotent.*
```
Comment (jkenny) [inline _posts/2016-06-24-restful-api-design-principles.md]: I would leave out the idempotent part here. Same for the 404 entry below.

File: _posts/2016-06-24-restful-api-design-principles.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+```
+{
+    "link": {
+        "rel" : "collection",
+        "href" : "https://api.example.com/zoos",
+        "type" : "application/vnd.example+json"
+    }
+}
+```
+Such a response means that there is a link that the client could follow to do more operations. The relationship between the link and the current response is defined by *"ref"* field, the URL of the resource this link points to is defined by *"href"* field, and the type of document returned by that link is defined by *"type"* field.
```
Comment (jkenny) [inline _posts/2016-06-24-restful-api-design-principles.md]: "The relationship between the link and the current response is defined by the *"rel"* field, ..."

**Comment by danielwpz** [comment]
Fixed errors @jkenny . 
And for the 'idempotent' part, I personally believe it is worth mentioning since some APIs don't distinguish between 201 and 200 and they return 200 after POST, which is not safe actually.

**Comment by shellscape** [comment]
@danielwpz is this dead?

**Pull request closed without merging**